paddle/fluid/framework/ir/mkldnn/params_to_int8_pass.cc

```diff
@@ -0,0 +1,246 @@
+    const int groups =
+        std::max(conv_op->Op()->GetAttrIfExists<int>("groups"), 1);
+
+    const auto& scale_weights_data =
+        conv_op->Op()->GetAttrIfExists<std::vector<float>>("Scale_weights");
+
+    bool is_multi_channel = scale_weights_data.size() > 1;
+
+    int count = 1;
+    if (is_multi_channel) {
+      count *= weights_tz[0];
+      if (groups > 1) {
+        count *= weights_tz[1];
+      }
+    }
+
+    // get scope to interact with tensors
+    auto* scope = param_scope();
+    PADDLE_ENFORCE_NOT_NULL(
+        scope, platform::errors::InvalidArgument("Scope cannot be nullptr."));
+
+    // get float weights
+    auto* weights =
+        scope->FindVar(conv_filter->Name())->GetMutable<LoDTensor>();
+
+    if (weights->dtype() != phi::DataType::FLOAT32) {
+      VLOG(4) << "Skipping convolution (id: " << conv_op->id()
+              << ") because it's a bug that it is detected again.";
+      return;
+    }
+
+    auto weights_shape = weights->dims();
+    float* weights_data = weights->data<float>();
+
+    // Create int8 weights variable
+    VarDesc int_weights(
+        patterns::PDNodeName(name_scope, "conv2d_int8_weights"));
+    int_weights.SetShape(weights_tz);
+    int_weights.SetDataType(framework::proto::VarType::Type::VarType_Type_INT8);
+    // int_weights.SetLoDLevel(conv_filter->Var()->GetLoDLevel());
+    int_weights.SetPersistable(true);
+    ir::Node* int_weights_node = g->CreateVarNode(&int_weights);
+    auto* int_weights_tensor =
+        scope->Var(int_weights_node->Name())->GetMutable<LoDTensor>();
+
+    // Initialize int8 weights
+    int_weights_tensor->Resize(weights_shape);
+    auto int_weights_data =
+        int_weights_tensor->mutable_data<int8_t>(platform::CPUPlace());
+    // Quantize weights to int8
+    auto len = weights->numel() / count;
+    if (is_multi_channel) {
+      for (int i = 0; i < count; ++i) {
+        auto scale = scale_weights_data[i];
+        int_weights_data =
+            std::transform(weights_data, weights_data + len, int_weights_data,
+                           [&scale](float f) {
+                             return static_cast<int8_t>(std::round(f * scale));
+                           });
+        weights_data += len;
+      }
+    } else {
+      auto len = int_weights_tensor->numel();
+      auto scale = scale_weights_data[0];
+      std::transform(weights_data, weights_data + len, int_weights_data,
+                     [&scale](float f) {
+                       return static_cast<int8_t>(std::round(f * scale));
+                     });
+    }
+
+    // connect new int8 tensor to weights input
+    conv_op->Op()->SetInput("Filter", {int_weights_node->Name()});
+    IR_NODE_LINK_TO(int_weights_node, conv_op);
+    GraphSafeRemoveNodes(graph, {conv_filter});
+    conv_op->Op()->SetAttr("Scale_weights", std::vector<float>(1, 1));
+
+    // Get float biases
+    if (has_bias && conv_op->Op()->Input("Bias").size() > 0) {
```

Comment:
I think you can merge data calculation and setting INT8 Bias/Weights and move it to separate function.